### PR TITLE
feat(masthead-v2.1): "view all" CTA design updates

### PR DESCRIPTION
**Pull request by jkaeser**
### Related Ticket(s)

https://jsw.ibm.com/browse/ADCMS-2818

### Description

Updates "view all" CTAs to comply with [the latest v2.1.0 design & functional specifications](https://www.figma.com/file/35JfN0Sg4ltIIjM7jBDKMk/Navigation?node-id=3179%3A173922&t=QbcD40zoyhOitQ5e-0).

### Changelog

**Added**

- Add data model support for choosing the megapanel column the "view all" CTA will appear in.

**Changed**

- Update "view all" CTA styles and markup to match v2.1.0 specifications.
- Prevent stylelint from blocking `::part` pseudo-class usage.

### To review

Test on https://ibmdotcom-webcomponents.s3.us-east.cloud-object-storage.appdomain.cloud/deploy-previews/9939/iframe.html?args=&knob-Card%20Heading%20(heading):=Explore%20AI%20use%20cases%20in%20all%20industries&knob-Card%20Href%20(href):=https://example.com&knob-Open%20(open)=true&knob-Button%20content=Lorem%20ipsum%20dolor&knob-platform%20name%20(platform)=&knob-show%20the%20profile%20functionality%20(has-profile)=true&knob-show%20the%20search%20functionality%20(has-search)=true&knob-search%20placeholder%20(searchPlaceholder)=Search%20all%20of%20IBM&knob-selected%20menu%20item%20(selected-menu-item)=Consulting%20&%20Services=&knob-The%20user%20authenticated%20status%20(user-status)=Unauthenticated&knob-use%20mock%20nav%20data%20(use-mock)=true&id=components-masthead--with-v-2-data&viewMode=story.

Verify "view all" CTAs appear correctly regardless of layout and column:

1. **Left column:** "Faceted Megamenu" link
2. **Center column:** "Simple Megamenu" link
3. **Right column:** "Simple Megamenu (Featured)" link

**Note:** this round of work does not include mobile "view all" CTAs.

### CODE REVIEW

**Comment by proeung** [comment]
Note: Adding myself and others to this PR when it's ready for review.

**Comment by andy-blum** [comment]
I could be misunderstanding how this all works, but I think you have the view alls mixed up between featured and faceted?

Faceted Design
Screenshot 2023-02-09 at 12 51 50 PM

Faceted Implementation
Screenshot 2023-02-09 at 12 52 13 PM

Featured Design
Screenshot 2023-02-09 at 12 52 01 PM

Featured Implementation
Screenshot 2023-02-09 at 12 52 30 PM

**Comment by jkaeser** [comment]
> I could be misunderstanding how this all works, but I think you have the view alls mixed up between featured and faceted?

A functional spec mandates it should be optional! You can have them appear in either column in either layout.

File: packages/web-components/src/components/masthead/masthead-composite.ts [hunk 1]
```diff
@@ -188,7 +191,11 @@ class DDSMastheadComposite extends HostListenerMixin(LitElement) {
           const { itemKey, groups, heading } = item;
           return html`
             <div id="panel-${itemKey}" role="tabpanel" aria-labelledby="tab-${itemKey}" hidden>
-              <dds-megamenu-right-navigation style-scheme="${MEGAMENU_RIGHT_NAVIGATION_STYLE_SCHEME.TAB}">
+              <dds-megamenu-right-navigation
+                style-scheme="${MEGAMENU_RIGHT_NAVIGATION_STYLE_SCHEME.TAB}"
+                view-all-href="${viewAll?.position !== MEGAPANEL_VIEW_ALL_POSITION.START ? ifNonNull(viewAll?.url) : ''}"
```
[the remark this answers is not in the document]
Comment (jkaeser) [reply packages/web-components/src/components/masthead/masthead-composite.ts:196]: My reasoning was if a user sets a `viewAll.url` in the incoming data, they probably intend the CTA to appear. This approach defaults to showing it _somewhere_ (on the right side) in the event `viewAll.position` is missing or has a typo. No CTA will appear if both `viewAll.url` and `viewAll.position` are missing.

That all being said, I can see why this behavior would be confusing. Still think it's wiser to do a strict equality check instead? I'm open to it.

File: packages/stylelint-config-ibmdotcom/rules/possible-errors.js [hunk 2]
```diff
@@ -42,8 +42,12 @@ module.exports = {
     'block-no-empty': true,
 
     // Selector
-    'selector-pseudo-class-no-unknown': true,
-    'selector-pseudo-element-no-unknown': true,
+    'selector-pseudo-class-no-unknown': [
+      true,
+      {
+        ignorePseudoClasses: ['part'],
```
[the remark this answers is not in the document]
Comment (jkaeser) [reply packages/stylelint-config-ibmdotcom/rules/possible-errors.js]: Agreed! Better late than never 🙂

File: packages/web-components/src/components/masthead/masthead-composite.ts [hunk 3]
```diff
@@ -171,7 +171,10 @@ class DDSMastheadComposite extends HostListenerMixin(LitElement) {
 
     return html`
       <dds-megamenu layout="${MEGAMENU_LAYOUT_SCHEME.TAB}">
-        <dds-megamenu-left-navigation view-all-href="${ifNonNull(viewAll?.url)}" view-all-title="${ifNonNull(viewAll?.title)}">
+        <dds-megamenu-left-navigation
+          view-all-href="${viewAll?.position === MEGAPANEL_VIEW_ALL_POSITION.START ? ifNonNull(viewAll?.url) : ''}"
```
[the remark this answers is not in the document]
Comment (jkaeser) [reply packages/web-components/src/components/masthead/masthead-composite.ts:175]: Working on getting logical properties to be second nature, but you've convinced me here. Makes a lot of sense considering the `right` and `left` navigation components.

File: packages/web-components/src/components/masthead/masthead-composite.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (andy-blum) [reply packages/web-components/src/components/masthead/masthead-composite.ts:196]: @proeung @jwitkowski79 @oliviaflory 

Any concerns on this defaulting behavior?

[the remark this answers is not in the document]
Comment (proeung) [reply packages/web-components/src/components/masthead/masthead-composite.ts:196]: I feel like the `start` position (see attached) is the most commonly used position within the megamenu panel based on some of the design iterations that I've seen, therefore the current behavior of defaulting to `start` makes sense to me. However, @jwitkowski79 @oliviaflory let me know if you agree.

Screen Shot 2023-02-13 at 11 14 52 AM

As for the data, we'll most impose a strict content entry process where the `position` and `url` will always be filled out if there's a CTA to be displayed within each menu panel.

```
viewAll: {
    title: 'View all',
    titleEnglish: 'View all',
    url: 'https://www.example.com',
    position: 'start',
      }
```

**Review by proeung — APPROVED** [review]
LGTM!

**Comment by proeung** [comment]
@jkaeser @andy-blum I'm good to merge the changes of this PR so that we can move forward with the merging/clean-up PR (https://github.com/carbon-design-system/carbon-for-ibm-dotcom/pull/10031). Give me a 👍 , if you both agree.

File: packages/web-components/src/components/masthead/masthead-composite.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (oliviaflory) [reply packages/web-components/src/components/masthead/masthead-composite.ts:196]: @proeung I believe it depends on the type of menu. I agree the `start` position is most common for the `standard` and _likely_ `featured sidebar` menus (though featured can be either), but in the faceted menu it is most commonly in the red outlined box.

**Comment by jkaeser** [comment]
> @jkaeser @andy-blum I'm good to merge the changes of this PR so that we can move forward with the merging/clean-up PR (#10031). Give me a 👍 , if you both agree.

I have changes incoming for some of the feedback. Going to use directional names instead of logical properties, and shifting the default position of the view all CTA to the left on tabbed/faceted layouts (to align to your expectation).

File: packages/web-components/src/components/masthead/masthead-composite.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (jkaeser) [reply packages/web-components/src/components/masthead/masthead-composite.ts]: Thanks, @oliviaflory!

Given that feedback, I need to make the CTA appear on the left by default for Featured layout. Standard doesn't have a multi-column megamenu layout and therefore doesn't get any options.

**Pull request merged**